### PR TITLE
switching to GeneralSFMFactor2 and adding option for shared intrinsics

### PR DESCRIPTION
**Pull request by ayushbaid**
This PR depends on https://github.com/borglab/gtsam/pull/644

### CODE REVIEW

**Comment by ayushbaid** [comment]
# Metrics

## Master
```
[2021-04-07 01:39:18,136 DEBUG data_assoc.py line 124 572875] [Data association] output number of tracks: 9111
[2021-04-07 01:39:18,137 DEBUG data_assoc.py line 125 572875] [Data association] output avg. track length: 5.25
[2021-04-07 01:39:18,432 INFO bundle_adjustment.py line 40 572875] Input: 9111 tracks on 12 cameras

[2021-04-07 01:39:24,399 INFO bundle_adjustment.py line 105 572875] initial error: 1187727.51
[2021-04-07 01:39:24,400 INFO bundle_adjustment.py line 106 572875] final error: 6434.54
[2021-04-07 01:39:31,107 INFO geometry_comparisons.py line 69 572874] Sim(3) Rotation `aRb`: rz=-161.03 deg., ry=23.03 deg., rx=23.20 deg.
[2021-04-07 01:39:31,109 INFO geometry_comparisons.py line 72 572874] Sim(3) Translation `atb`: [tx,ty,tz]=[-0.92 -0.    0.17]
[2021-04-07 01:39:31,110 INFO geometry_comparisons.py line 73 572874] Sim(3) Scale `asb`: 0.86
[2021-04-07 01:39:31,112 INFO geometry_comparisons.py line 81 572874] Pose graph Sim(3) alignment complete.
[2021-04-07 01:39:31,119 INFO geometry_comparisons.py line 69 572874] Sim(3) Rotation `aRb`: rz=-160.99 deg., ry=23.01 deg., rx=23.14 deg.
[2021-04-07 01:39:31,119 INFO geometry_comparisons.py line 72 572874] Sim(3) Translation `atb`: [tx,ty,tz]=[-0.98  0.    0.22]
[2021-04-07 01:39:31,119 INFO geometry_comparisons.py line 73 572874] Sim(3) Scale `asb`: 0.84
[2021-04-07 01:39:31,119 INFO geometry_comparisons.py line 81 572874] Pose graph Sim(3) alignment complete.
[2021-04-07 01:40:01,878 INFO run_scene_optimizer.py line 46 572866] Scene avg reproj error: 0.189
```

## SfmFactor2 w/o shared intrinsics:
```
[2021-04-07 01:42:52,676 DEBUG data_assoc.py line 124 573402] [Data association] output number of tracks: 9144
[2021-04-07 01:42:52,676 DEBUG data_assoc.py line 125 573402] [Data association] output avg. track length: 5.24
[2021-04-07 01:42:52,968 INFO bundle_adjustment.py line 125 573402] Input: 9144 tracks on 12 cameras

[2021-04-07 01:42:59,964 INFO bundle_adjustment.py line 165 573402] initial error: 3407519.55
[2021-04-07 01:42:59,966 INFO bundle_adjustment.py line 166 573402] final error: 3911.85
[2021-04-07 01:43:06,769 INFO geometry_comparisons.py line 69 573403] Sim(3) Rotation `aRb`: rz=114.21 deg., ry=-17.41 deg., rx=-59.14 deg.
[2021-04-07 01:43:06,770 INFO geometry_comparisons.py line 72 573403] Sim(3) Translation `atb`: [tx,ty,tz]=[-0.93 -0.01  0.17]
[2021-04-07 01:43:06,770 INFO geometry_comparisons.py line 73 573403] Sim(3) Scale `asb`: 0.75
[2021-04-07 01:43:06,771 INFO geometry_comparisons.py line 81 573403] Pose graph Sim(3) alignment complete.
[2021-04-07 01:43:06,773 INFO geometry_comparisons.py line 69 573403] Sim(3) Rotation `aRb`: rz=114.72 deg., ry=-17.15 deg., rx=-58.96 deg.
[2021-04-07 01:43:06,773 INFO geometry_comparisons.py line 72 573403] Sim(3) Translation `atb`: [tx,ty,tz]=[-0.94 -0.    0.12]
[2021-04-07 01:43:06,789 INFO geometry_comparisons.py line 73 573403] Sim(3) Scale `asb`: 0.75
[2021-04-07 01:43:06,789 INFO geometry_comparisons.py line 81 573403] Pose graph Sim(3) alignment complete.
[2021-04-07 01:43:37,318 INFO run_scene_optimizer.py line 46 573392] Scene avg reproj error: 0.192
```

## SfmFactor2 w/ shared intrinsics:
```
[2021-04-07 01:45:37,262 DEBUG data_assoc.py line 124 573721] [Data association] output number of tracks: 9024
[2021-04-07 01:45:37,262 DEBUG data_assoc.py line 125 573721] [Data association] output avg. track length: 5.19
[2021-04-07 01:45:37,551 INFO bundle_adjustment.py line 125 573721] Input: 9024 tracks on 12 cameras

[2021-04-07 01:45:44,744 INFO bundle_adjustment.py line 165 573721] initial error: 3013579.99
[2021-04-07 01:45:44,754 INFO bundle_adjustment.py line 166 573721] final error: 4812.57
[2021-04-07 01:45:52,335 INFO geometry_comparisons.py line 69 573723] Sim(3) Rotation `aRb`: rz=-127.39 deg., ry=60.49 deg., rx=-115.51 deg.
[2021-04-07 01:45:52,335 INFO geometry_comparisons.py line 72 573723] Sim(3) Translation `atb`: [tx,ty,tz]=[-1.   -0.    0.17]
[2021-04-07 01:45:52,335 INFO geometry_comparisons.py line 73 573723] Sim(3) Scale `asb`: 0.75
[2021-04-07 01:45:52,335 INFO geometry_comparisons.py line 81 573723] Pose graph Sim(3) alignment complete.
[2021-04-07 01:45:52,341 INFO geometry_comparisons.py line 69 573723] Sim(3) Rotation `aRb`: rz=-126.35 deg., ry=60.25 deg., rx=-116.38 deg.
[2021-04-07 01:45:52,343 INFO geometry_comparisons.py line 72 573723] Sim(3) Translation `atb`: [tx,ty,tz]=[-1.01 -0.    0.12]
[2021-04-07 01:45:52,343 INFO geometry_comparisons.py line 73 573723] Sim(3) Scale `asb`: 0.75
[2021-04-07 01:45:52,343 INFO geometry_comparisons.py line 81 573723] Pose graph Sim(3) alignment complete.
[2021-04-07 01:46:20,185 INFO run_scene_optimizer.py line 46 573712] Scene avg reproj error: 0.2
```

File: gtsfm/bundle/bundle_adjustment.py [hunk 1]
```diff
@@ -28,6 +41,78 @@ class BundleAdjustmentOptimizer:
     This class refines global pose estimates and intrinsics of cameras, and also refines 3D point cloud structure given
     tracks from triangulation."""
 
+    def __init__(self, shared_calib: bool = False):
+        """Initializes the parameters for bundle adjustment module.
+
+        Args:
+            shared_calib (optional): Flag to enable shared calibration across
```
Comment (johnwlambert) [inline gtsfm/bundle/bundle_adjustment.py:48]: shared_calib should go in the config i believe. COLMAP allows this as a config parameter

Comment (johnwlambert) [reply gtsfm/bundle/bundle_adjustment.py:48]: oh nevermind, i saw you added that

File: gtsfm/bundle/bundle_adjustment.py [hunk 2]
```diff
@@ -28,6 +41,78 @@ class BundleAdjustmentOptimizer:
...
+                                     all cameras. Defaults to False.
+        """
+
+        self._shared_calib = shared_calib
+
+    def __add_camera_prior_and_initial_value(
```
Comment (johnwlambert) [inline gtsfm/bundle/bundle_adjustment.py:54]: maybe should be two separate functions?

`def __add_pose_prior()`
`def __initialize()`

File: gtsfm/bundle/bundle_adjustment.py [hunk 3]
```diff
@@ -42,71 +127,46 @@ def run(self, initial_data: GtsfmData) -> SfmResult:
...
-                C(valid_camera_indices[0]),
-                initial_data.get_camera(valid_camera_indices[0]),
-                gtsam.noiseModel.Isotropic.Sigma(PINHOLE_CAM_CAL3BUNDLER_DOF, 0.1),
-            )
-        )
-        # Also add a prior on the position of the first landmark to fix the scale
-        graph.push_back(
-            gtsam.PriorFactorPoint3(
-                P(0), initial_data.get_track(0).point3(), gtsam.noiseModel.Isotropic.Sigma(POINT3_DOF, 0.1),
+            self.__add_measurement_factors_and_initial_values(
```
Comment (johnwlambert) [inline gtsfm/bundle/bundle_adjustment.py:140]: is this with black line length 120?

File: gtsfm/bundle/bundle_adjustment.py [hunk 4]
```diff
@@ -42,71 +127,46 @@ def run(self, initial_data: GtsfmData) -> SfmResult:
...
+                graph, initial_values, initial_data.get_track(j), j, measurement_noise,
```
Comment (johnwlambert) [inline gtsfm/bundle/bundle_adjustment.py:141]: maybe its cleaner to just do:
```python
for j in range(initial_data.number_tracks()):
        self.__add_measurement_factors(graph, initial_values, initial_data.get_track(j), j)
        # add initial value for 3d point
        initial_values.insert(P(track_idx), track.point3())
```

File: gtsfm/bundle/bundle_adjustment.py [hunk 5]
```diff
@@ -9,13 +9,15 @@
...
 
 from gtsfm.common.sfm_result import SfmResult
 
 # TODO: any way this goes away?
-C = symbol_shorthand.C
-P = symbol_shorthand.P
+C = symbol_shorthand.C  # camera
+P = symbol_shorthand.P  # 3d point
+X = symbol_shorthand.X  # camera pose
+K = symbol_shorthand.K  # calibration
```
[the remark this answers is not in the document]
Comment (johnwlambert) [reply gtsfm/bundle/bundle_adjustment.py:28]: I would like to include the info still, to help users understand why we use the GeneralSfmFactor2 instead of GeneralSfmFactor (and maybe point to URL documentation of each -- and how factors differe mathematically)

File: gtsfm/bundle/bundle_adjustment.py [hunk 6]
```diff
@@ -28,6 +41,78 @@ class BundleAdjustmentOptimizer:
...
+        Args:
+            graph: factor graph for the problem.
+            initial_values: object holding the initial values for camera
+                            parameters and 3d points.
+            camera: the camera object to add to graph and to initial values.
+            camera_idx: index of the camera.
+        """
+
+        # add prior factor for pose
+        graph.push_back(PriorFactorPose3(X(camera_idx), camera.pose(), Isotropic.Sigma(CAM_POSE3_DOF, 0.1),))
```
Comment (johnwlambert) [inline gtsfm/bundle/bundle_adjustment.py:69]: why are we putting a prior factor on every pose3?

before it was just on the first pose

File: gtsfm/bundle/bundle_adjustment.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (ayushbaid) [reply gtsfm/bundle/bundle_adjustment.py:38]: added a small note on why we shift. Mathematically, the factors are the same.

**Comment by johnwlambert** [comment]
Hi @ayushbaid, please revisit and resolve conflicts

File: gtsfm/bundle/bundle_adjustment.py [hunk 7]
```diff
@@ -5,37 +5,147 @@
...
+            graph: factor graph for the problem.
+            camera: the camera object to add to graph and to initial values.
+            camera_idx: index of the camera.
+        """
+        graph.push_back(
+            PriorFactorPose3(X(camera_idx), camera.pose(), Isotropic.Sigma(CAM_POSE3_DOF, CAM_POSE3_PRIOR_NOISE_SIGMA))
+        )
+
+        # process just one camera if shared calibration; otherwise process all
+        if camera_idx == 0 or not self._shared_calib:
```
Comment (ayushbaid) [inline gtsfm/bundle/bundle_adjustment.py:90]: maybe needs a flag to tell if a calibration prior has been added instead of check on index.

File: gtsfm/bundle/bundle_adjustment.py (hunk 6 above)
Comment (ayushbaid) [reply gtsfm/bundle/bundle_adjustment.py]: I remember Frank mentioning that we should use priors on camera poses because we have good initializations from the averaging module. Maybe this needs to be tested first in a separate PR?

File: gtsfm/bundle/bundle_adjustment.py (hunk 7 above)
Comment (johnwlambert) [reply gtsfm/bundle/bundle_adjustment.py:90]: i think this check should come on the caller side, not inside the subroutine. otherwise it can confuse people, thinking we add a prior for every camera @ayushbaid

File: gtsfm/bundle/bundle_adjustment.py [hunk 8]
```diff
@@ -5,37 +5,147 @@
...
 METRICS_PATH = Path(__file__).resolve().parent.parent.parent / "result_metrics"
 
+"""In this file, we use the GTSAM's GeneralSFMFactor2 instead of GeneralSFMFactor because Factor2 enables decoupling
+of the camera pose and the camera intrinsics, and hence gives an option to share the intrinsics between cameras.
+"""
+
 # TODO: any way this goes away?
-C = symbol_shorthand.C
-P = symbol_shorthand.P
+P = symbol_shorthand.P  # 3d point
```
Comment (johnwlambert) [inline gtsfm/bundle/bundle_adjustment.py:37]: just curious, why the change in variable name here from C to X?

File: gtsfm/bundle/bundle_adjustment.py [hunk 9]
```diff
@@ -58,71 +170,47 @@ def run(self, initial_data: GtsfmData) -> Tuple[GtsfmData, GtsfmMetricsGroup]:
...
-            initial.insert(P(j), track.point3())
+            self.__add_measurement_factor(graph, track, j, measurement_noise)
+            initial_values.insert(P(j), track.point3())
+
+        # add prior and initial values for cameras
+        for i in initial_data.get_valid_camera_indices():
+            initialized_cam = initial_data.get_camera(i)
+
+            self.__add_camera_prior(graph, initialized_cam, i)
+            self.__add_camera_initial_value(initial_values, initialized_cam, i)
```
Comment (johnwlambert) [inline gtsfm/bundle/bundle_adjustment.py:192]: i actually find this helper a bit less readable than before. since basically we are wrapping a 1-liner. and i think the 1-liner is clearer to me than the new method

File: gtsfm/bundle/bundle_adjustment.py (hunk 8 above)
Comment (ayushbaid) [reply gtsfm/bundle/bundle_adjustment.py:36]: C is for camera. I thought of using X for pose, as http://www.cs.cmu.edu/~kaess/pub/Dellaert17fnt.pdf uses *X* to describe robot locations [Fig 1.1]

**Comment by johnwlambert** [comment]
Thanks for the PR, Ayush. So it looks like pose errors after BA go down

9->7 in rot angle
0.52 -> 0.4 in trans dist
8.8 -> 7.4 in trans angle

but we go from 2236 to just 80 tracks. Any idea why that occurs?

Looks promising but I think we would need to do more digging.

**Comment by ayushbaid** [comment]
Even the `mean_reprojection_errors_filtered_px ` is Nan. Something is broken I think.

**Comment by ayushbaid** [comment]
Significant improvements after essentially fixing the calibrations during BA optimization.

# Door-12 SIFT front end

## Baseline


Metric name | Baseline | New
-- | -- | -- 
number_cameras | 12 | 12
mean_rotation_error_angle_deg | 0.121735 |  0.023079 
median_rotation_error_angle_deg | 0.127 | 0.014
mean_translation_error_distance | 0.0505733 |  0.00507481
median_translation_error_distance | 0.042 | 0.005
mean_translation_angle_error_deg | 1.594 |  0.0815243 
median_translation_angle_error_deg | 0.986 | 0.052
number_tracks_filtered | 8723 | 8989
mean_3d_track_lengths_filtered | 5.26826 |  4.83613

**Comment by ayushbaid** [comment]
> Thanks for the PR, Ayush. So it looks like pose errors after BA go down
> 
> 9->7 in rot angle
> 0.52 -> 0.4 in trans dist
> 8.8 -> 7.4 in trans angle
> 
> but we go from 2236 to just 80 tracks. Any idea why that occurs?
> 
> Looks promising but I think we would need to do more digging.

@johnwlambert I have a theory of what goes wrong: In the baseline, our calibrations have fewer constraints and can take weird values to get to a local minima. However, with the introduction of additional constraints, we lose that DoF which allows us to reduce the overall error but leads to a 'bad' geometric solution. 

I have now essentially fixed the calibrations during DA. Let's see the results with them.

**Comment by ayushbaid** [comment]
# Notredame



Metric name | Baseline | New
-- | -- | -- 
number_cameras | 19 | 19
number_tracks_unfiltered | 2366 | 2366
mean_rotation_error_angle_deg | 0.659341 |  5.64764 
mean_translation_error_distance | 0.253517 |  3.13774 
mean_translation_angle_error_deg | 4.99404 |  38.2404 
number_tracks_filtered | 2075 | 1513
mean_3d_track_lengths_filtered | 2.65012 |  2.24719 
mean_reprojection_errors_filtered_px | 0.789671 |  0.751667

Every metric goes down. Even the median values.

**Comment by ayushbaid** [comment]
# Skydio-32



Metric name | Baseline | New
-- | -- | -- 
number_cameras | 32 | 32
number_tracks_unfiltered | 5567 | 5529
mean_rotation_error_angle_deg | 8.02652 |  2.69251
mean_translation_error_distance |  0.688316 |  0.481073 
mean_translation_angle_error_deg |  10.4483  |  4.75131
number_tracks_filtered | 1256  | 3378
mean_3d_track_lengths_filtered |  2.06369  |  2.32268
mean_reprojection_errors_filtered_px |  1.41338  |  0.825517

Improvements everywhere, even in medians.

**Comment by ayushbaid** [comment]
# Palace of fine arts (SIFT)



Metric name | Baseline | New
-- | -- | -- 
number_cameras | 281 | 281
number_tracks_unfiltered | 8415 | 8507 
mean_rotation_error_angle_deg | 17.7185  |   17.7224 
mean_translation_error_distance |  6.39294 |   7.97013
mean_translation_angle_error_deg |   27.0748  |   24.084 
number_tracks_filtered | 2673  | 4008 
mean_3d_track_lengths_filtered |  2.03966  |  2.26397
mean_reprojection_errors_filtered_px |  1.32607  | 1.00759

Improvements in all metrics except the translation error distance.

File: gtsfm/bundle/bundle_adjustment.py [hunk 10]
```diff
@@ -91,19 +114,48 @@ def run(
...
+            PriorFactorPose3(
+                X(valid_camera_indices[0]),
+                initial_data.get_camera(valid_camera_indices[0]).pose(),
+                gtsam.noiseModel.Isotropic.Sigma(CAM_POSE3_DOF, CAM_POSE3_PRIOR_NOISE_SIGMA),
             )
         )
+
+        # add prior on all calibrations
+        if self._shared_calib:
+            i = valid_camera_indices[0]
```
Comment (akshay-krishnan) [inline gtsfm/bundle/bundle_adjustment.py:141]: just have a new variable: 
camera_indices_with_prior = [0] if shared else valid_camera_indices. 

Then use the for loop with camera_indices_with_prior :)

File: gtsfm/common/gtsfm_data.py [hunk 11]
```diff
@@ -306,10 +306,18 @@ def get_avg_scene_reprojection_error(self) -> float:
...
+        )
+        logger.info(
+            "Avg scene reproj error: %.3f", np.nanmean(scene_reproj_errors) if len(scene_reproj_errors) else np.NaN
+        )
+        logger.info(
+            "Median scene reproj error: %.3f", np.nanmedian(scene_reproj_errors) if len(scene_reproj_errors) else np.NaN
+        )
+        logger.info(
+            "Max scene reproj error: %.3f", np.nanmax(scene_reproj_errors) if len(scene_reproj_errors) else np.NaN
+        )
```
Comment (akshay-krishnan) [inline gtsfm/common/gtsfm_data.py:320]:
```
If len(scene_reproj_errors) is 0: 
   log.error("log_scene_reprojection_error_stats: Scene projection errors is empty!")
   return 
log as usual
```

**Comment by ayushbaid** [comment]
Please refer to this for CI: https://github.com/borglab/gtsfm/actions/runs/1276212179

File: gtsfm/common/gtsfm_data.py (hunk 11 above)
Comment (ayushbaid) [reply gtsfm/common/gtsfm_data.py:320]: thanks. made the change

File: gtsfm/bundle/bundle_adjustment.py (hunk 10 above)
Comment (ayushbaid) [reply gtsfm/bundle/bundle_adjustment.py]: I implemented a similar thing. Please let me know if it works.

**Comment by johnwlambert** [comment]
Translation error distance and mean translation error anguler error go down by 10x on  Door-12, deep-front-end

Bundle Adjustment Metrics

| Metric name | 	Value Before | Value After | 
| :-: | :-: | :-: | 
| number_cameras |	12 | 12 |
| number_tracks_unfiltered |	5412 | 5418  |
| mean_3d_track_lengths_unfiltered |	6.91981 | 6.91713 | 
| mean_reprojection_errors_unfiltered_px |	0.770565 | 0.771594 | 
| mean_rotation_error_angle_deg |	0.036867 | 0.0205298 | 
| mean_translation_error_distance |	0.0203584 | 0.00226694 | 
| mean_translation_angle_error_deg |	0.773877 | 0.074895 | 
| number_tracks_filtered |	5118 | 5034 | 
| mean_3d_track_lengths_filtered |	6.85678 | 6.84684 | 
| mean_reprojection_errors_filtered_px |	0.603743 | 0.613359 |

**Comment by johnwlambert** [comment]
on skydio-32 w/ deep front end, mean_translation_error_distance seems to be halved, mean_translation_angle_error_deg from 17->4, and mean reproj error 1.6 -> 0.8
Bundle Adjustment Metrics

| Metric name |	Value Before | Value After | 
| :-: | :-: | :-: |
| number_cameras |	32 | 32
| number_tracks_unfiltered |	6172 | 5563
| mean_3d_track_lengths_unfiltered |	2.78905 | 2.75481
| mean_reprojection_errors_unfiltered_px |	31.9467 | 6.21121
| mean_rotation_error_angle_deg |	7.86718 | 2.51154
| mean_translation_error_distance |	1.13207 | 0.402831
| mean_translation_angle_error_deg |	17.2021 | 4.27734
| number_tracks_filtered |	638 | 3417
| mean_3d_track_lengths_filtered |	2.0768  | 2.31255
| mean_reprojection_errors_filtered_px  |	1.60276 | 0.820945 |

**Comment by johnwlambert** [comment]
on AstronNet RC w/ sift front-end, translation error distance doubles, which doesnt seem good, but we get 10x as many good tracks. Maybe the regression is random, I'll check the logs on the second CI run:

Bundle Adjustment Metrics

|Metric name |	Value Before | Value After  |
| :-: | :-: | :-: |
|number_cameras |	63 | 62  |
|number_tracks_unfiltered |	11834 | 10372  |
|mean_3d_track_lengths_unfiltered |	2.11653 | 2.12669  |
|mean_reprojection_errors_unfiltered_px |	38.3841 | 1.06983  |
|mean_rotation_error_angle_deg |	7.53114 | 7.387  |
|mean_translation_error_distance |	2295.06 | 5354.17  |
|mean_translation_angle_error_deg |	16.2376 | 16.1033  |
|number_tracks_filtered |	919 | 9616  |
|mean_3d_track_lengths_filtered |	2 | 2.09214  |
|mean_reprojection_errors_filtered_px |	1.72802 | 0.339136  |

**Comment by johnwlambert** [comment]
On Palace of Fine Arts, w/ the sift front end, we get twice as many good tracks, with reproj error decreasing from 1.2 -> 0.8. All metrics are improved here:

Bundle Adjustment Metrics

| Metric name |	Value Before | Value After | 
| :-: | :-: | :-: |
| number_cameras |	281 | 281 | 
| number_tracks_unfiltered |	8297 | 8387 | 
| mean_3d_track_lengths_unfiltered |	3.11631 | 3.11124 | 
| mean_reprojection_errors_unfiltered_px |	18.8402 | 7.05629 | 
| mean_rotation_error_angle_deg |	17.718 | 16.6419 | 
| mean_translation_error_distance |	8.0062 | 8.00944 | 
| mean_translation_angle_error_deg |	26.2418 | 24.2808 | 
| number_tracks_filtered |	2639 | 4593 | | 
| mean_3d_track_lengths_filtered |	2.02577 | 2.44807 | 
| mean_reprojection_errors_filtered_px |	1.28845 | 0.896363 |

**Comment by johnwlambert** [comment]
skydio-8, deep front end shows improvement in all metrics

Bundle Adjustment Metrics

| Metric name |	Value Before |  Value After | 
| :-: | :-: | :-: | 
| number_cameras |	8 | 8 |
| number_tracks_unfiltered |	2113 | 2113 |
| mean_3d_track_lengths_unfiltered |	2.87033 | 2.87033 |
| mean_reprojection_errors_unfiltered_px |	4.56096 | 1.83125 |
| mean_rotation_error_angle_deg |	0.858085 | 0.51455 |
| mean_translation_error_distance |	0.128875 | 0.0452944 |
| mean_translation_angle_error_deg |	1.55323 | 0.526738 |
| number_tracks_filtered |	1516 | 1747 |
| mean_3d_track_lengths_filtered |	2.60818 | 2.66171 |
| mean_reprojection_errors_filtered_px |	1.12901 | 0.840985 |

File: gtsfm/common/gtsfm_data.py (hunk 11 above)
Comment (johnwlambert) [reply gtsfm/common/gtsfm_data.py:320]: @ayushbaid why would there be a case when the scene_reproj_errors would be empty or None? have we ever seen that? that would mean we have zero optimized points after filtering, which I have never seen

**Review by johnwlambert — APPROVED** [review]
LGTM, nice work Ayush
